std/string.d

```diff
@@ -3215,6 +3186,202 @@ S center(S)(S s, size_t width, dchar fillChar = ' ') @trusted pure
+    else static if (C.sizeof == 4 && hasLength!Range)
+    {
+        import std.range : chain, repeat;
+
+        auto len = r.length;
+        if (len > width)
+            len = width;
+        const nleft = (width - len) / 2;
+        const nright = width - len - nleft;
+        return chain(repeat(fillChar, nleft), r, repeat(fillChar, nright));
+    }
+    else static if (C.sizeof == 4)
+    {
+        static struct Result
+        {
+          private:
+            Range _input;
+            size_t _width;
+            alias _width nleft;       // number of fill characters to prepend
+            size_t nright;            // number of fill characters to append
+            dchar _fillChar;
+            bool inited;
+
+            // Lazy initialization so constructor is trivial and cannot fail
+            void initialize()
+            {
+                // Replace _width with nleft
+                // (use alias instead of union because CTFE cannot deal with unions)
+                assert(_width);
+                // Lookahead to see now many fill characters are needed
+                size_t len = 0;
+                for (auto r = _input.save; true; r.popFront())
+                {
+                    if (len == _width)
+                    {
+                        nleft = 0;
+                        nright = 0;
+                        break;
+                    }
+                    if (r.empty)
+                    {
+                        auto n = (_width - len) / 2;
+                        nright = _width - len - n;
+                        nleft = n;
+                        break;
+                    }
+                    ++len;
+                }
+                inited = true;
+            }
+
+          public:
+            this(Range input, size_t width, dchar fillChar) pure nothrow
+            {
+                _input = input;
+                _fillChar = fillChar;
+                _width = width;
+            }
+
+            @property bool empty()
+            {
+                return !nleft && _input.empty && !nright;
+            }
+
+            @property C front()
+            {
+                C c;
+                if (!nleft)
+                {
+                    c = !_input.empty ? _input.front : _fillChar;   // fast path
+                }
+                else
+                {
+                    if (!inited)
+                        initialize();
+                    c = (nleft || _input.empty) ? _fillChar : _input.front;
+                }
+                return c;
+            }
+
+            void popFront()
+            {
+                if (!nleft)
+                {
+                    if (!_input.empty)
+                        _input.popFront();  // fast path
+                    else
+                        --nright;
+                }
+                else
+                {
+                    if (!inited)
+                        initialize();
+                    if (nleft)
```

Comment:
No, `initialize()` might change the state. This is exactly why I like using ranges much more than writing them.